source/grid/grid_tools_cache.cc

```diff
@@ -30,17 +30,47 @@ namespace GridTools
     , tria(&tria)
     , mapping(&mapping)
   {
-    tria_signal =
+    tria_change_signal =
       tria.signals.any_change.connect([&]() { mark_for_update(update_all); });
   }
 
+
+
+  template <int dim, int spacedim>
+  Cache<dim, spacedim>::Cache(const Triangulation<dim, spacedim> &tria)
+    : update_flags(update_all)
+    , tria(&tria)
+  {
+    tria_change_signal =
+      tria.signals.any_change.connect([&]() { mark_for_update(update_all); });
+
+    // Allow users to set this class up with an empty Triangulation and no
+    // Mapping argument by deferring Mapping assignment until after the
+    // Triangulation exists.
+    if (tria.get_reference_cells().size() == 0)
+      {
+        tria_create_signal = tria.signals.create.connect([&]() {
+          Assert(tria.get_reference_cells().size() > 0, ExcInternalError());
+          Assert(tria.get_reference_cells().size() == 1, ExcNotImplemented());
+          mapping = &tria.get_reference_cells()[0]
+                       .template get_default_linear_mapping<dim, spacedim>();
+        });
```

Comment:
My goal with this PR is to make the default constructor work with simplices, which means we need some way to set up the correct mapping based on the Triangulation's reference cell. Since the Triangulation might not have any cells at this point, e.g., we might have something like
```cpp
Triangulation<dim> tria;
GridCache<dim> cache(tria);
DoFHandler<dim> dof_handler(tria);
// etc
```
(this behavior is expected in at least one test) so we have to set up the mapping after `create_triangulation()` is called - hence the signal.